### PR TITLE
ci: Ensure all version references are up-to-date prior to release

### LINKED ISSUE #370: Ensure that README.md documentation states the latest release
See https://github.com/slsa-framework/slsa-verifier/pull/369

to ensure that the documentation is up-to-date with the latest release in SHA256SUM.md and the latest GitHub releases

### PR DESCRIPTION
**Pull request by pnacht**
Closes #370. Please look carefully, the workflow introduced here works on PRs, so make sure it doesn't allow untrusted content to be interpreted.

This PR adds a new pre-release check: in PRs with `#[]label:release` (adding dummy brackets not to trigger anything!) in the description, ensures that all version references are up-to-date. These include:

- the major version declared in the first line of `go.mod`;
- the most recent version declared in the first line of `SHA256SUM.md` (in both the link and the text);
- all references to slsa-verifier in the `README.md`.

Should any inconsistencies be found, the logs will display them, for example:

```
SHA256SUM.md and go.mod have different major versions:
go.mod:       v3
SHA256SUM.md: v2 (v2.0.1)
```

```
SHA256SUM.md: Visible text and linked URL do not match:
### [v2.0.1](https://github.com/slsa-framework/slsa-verifier/releases/tag/v2.0.2)
      ^^^^^                                                                ^^^^^ 
```

```
README.md and SHA256SUM.md refer to different versions:
SHA256SUM.md: v2.0.2
README.md:
  133:$ cd slsa-verifier && git checkout v2.0.1
  139:Download the binary from the latest release at [https://github.com/slsa-framework/slsa-verifier/releases/tag/v2.0.1
  139:](https://github.com/slsa-framework/slsa-verifier/releases/tag/v2.0.1
```

The logic for the README is brittle, relying on the fact that all current version references there are prefixed by the string `slsa-verifier`. Should this change in the future, the logic may need to change.

Also, this logic currently runs on all PRs marked with `#[]label:release`. Should release-candidate PRs be omitted?

### CODE REVIEW

File: .github/workflows/scripts/pre-release/references.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+version="$version_txt"
+
+# Ensure major version matches go.mod
+major_version_sha="$(sed -E 's/(.+)\..+\..+/\1/' <<< "$version")"
+
+if [[ "$major_version_go" != "$major_version_sha" ]]; then
+    echo "SHA256SUM.md and go.mod have different major versions:"
+    echo "go.mod:       v$major_version_go"
+    echo "SHA256SUM.md: v$major_version_sha (v$version)"
```
Comment (ianlewis) [inline .github/workflows/scripts/pre-release/references.sh:43]: `major_version_sha` makes it sound like it's a hash value.

```suggestion
major_version_md="$(sed -E 's/(.+)\..+\..+/\1/' <<< "$version")"

if [[ "$major_version_go" != "$major_version_md" ]]; then
    echo "SHA256SUM.md and go.mod have different major versions:"
    echo "go.mod:       v$major_version_go"
    echo "SHA256SUM.md: v$major_version_md (v$version)"
```

File: .github/workflows/scripts/pre-release/references.sh [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+### SHA256SUM.md
+###
+
+read -r line < SHA256SUM.md
+
+# Ensure both visible text and link point to the same release
+version_txt="$(sed -E "s~.*\[v(.*)\].*~\1~" <<< "$line")"
+version_lnk="$(sed -E "s~.*/v(.*)\)$~\1~" <<< "$line")"
+
+if [[ "$version_txt" != "$version_lnk" ]]; then
```
[the remark this answers is not in the document]
Comment (pnacht) [reply .github/workflows/scripts/pre-release/references.sh:24]: True. I now fetch the first non-blank line for both SHA256SUM.md and go.mod.

File: .github/workflows/scripts/pre-release/references.sh [hunk 3]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+# Verify that all references point to the same version
+
+set -euo pipefail
+
+# Get major version from go.mod
+major_version_go="$(head -n 1 go.mod | sed -E 's~.*/v(.*)~\1~')"
```
[the remark this answers is not in the document]
Comment (pnacht) [reply .github/workflows/scripts/pre-release/references.sh]: Done.

File: .github/workflows/scripts/pre-release/references.sh (hunk 1 above)
Comment (pnacht) [reply .github/workflows/scripts/pre-release/references.sh]: Fixed.

**Comment by pnacht** [comment]
An issue: I just noticed that I can't find any PRs that adopt the `#label:release` description. Is this actually used in this project or should I change the logic somehow?

**Comment by laurentsimon** [comment]
> An issue: I just noticed that I can't find any PRs that adopt the `#label:release` description. Is this actually used in this project or should I change the logic somehow?

e might not have released a verifier version yet with it, I don't remember.
Here's one example on the generator https://github.com/slsa-framework/slsa-github-generator/pull/1498

**Comment by ianlewis** [comment]
> > An issue: I just noticed that I can't find any PRs that adopt the `#label:release` description. Is this actually used in this project or should I change the logic somehow?
> 
> We might not have released a verifier version yet with it, I don't remember. Here's one example on the generator [slsa-framework/slsa-github-generator#1498](https://github.com/slsa-framework/slsa-github-generator/pull/1498)

Right, we may want to add it to the `RELEASE.md` docs as well if it's needed.

**Comment by pnacht** [comment]
I've just pushed a commit adding that instruction to RELEASE.md. I'm actually going to make a change to the script to take advantage of the fact that the version is actually declared in the `#label:release X.Y.Z`.

**Review by asraa — APPROVED** [review]
This looks great!

One qq -- can you fix DCO? (Or maybe TBH we should just remove it..)

**Comment by pnacht** [comment]
Ops, hadn't noticed the DCO. Fixed. I also just submitted a new commit to use the `#label:release v1.2.3` as the source-of-truth for the release version.

**Comment by ianlewis** [comment]
> This looks great!
> 
> One qq -- can you fix DCO? (Or maybe TBH we should just remove it..)

We may need to require it at some point (right now it's non-blocking). The slsa repo requires it for example.

**Pull request merged**